### PR TITLE
Define protocol buffers for gRPC Client backend

### PR DESCRIPTION
**Pull request by michaelmcallister**
This defines the initial protobuf contract for the gRPC Client as proposed in [RFD 56](https://github.com/gravitational/teleport/commit/f13e5312238a850e54a97e03c9ba56c84f8b4a7a)

The boilerplate for the implementation is inplace, with methods to be populated in subsequent PRs.

### CODE REVIEW

File: lib/backend/grpcclient/grpc.go
Comment (zmb3) [inline lib/backend/grpcclient/grpc.go:28]:
```suggestion
type Backend struct {
```

`grpcclient.GrpcBackend` says GRPC a lot. How about this naming instead?

File: lib/backend/grpcclient/grpc.go
Comment (zmb3) [inline lib/backend/grpcclient/grpc.go:37]:
```suggestion
// GetName returns the name of the backend as it appears in 'storage/type' section
```

File: lib/backend/grpcclient/grpc.go
Comment (zmb3) [inline lib/backend/grpcclient/grpc.go:34]: What's your plan for the actual implementation here?

**Review by zmb3 — COMMENTED** [review]
This looks fine, but I'm still a little unsure where we're headed with this work.

It feels like the motivation for this was a limitation with our Dynamo backend, and rather than addressing that limitation we're just adding another layer of abstraction between the Teleport code and the persistence layer.

How will the GRPC backend persist its data?

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:28]: Makes sense, thank you!

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:34]: The following configuration will be needed at the very minimum:

- gRPC server address
- serverCA
- ClientCert
- ClientKey

I can add this in now if you'd prefer, I don't mind

**Comment by michaelmcallister** [comment]
> It feels like the motivation for this was a limitation with our Dynamo backend, and rather than addressing that limitation we're just adding another layer of abstraction between the Teleport code and the persistence layer.

It's true that the main motivation is to address limitations with multi-tenancy and the current implementation of backends, but it's not necessarily DynamoDB specific. Whilst DynamoDB has unique constraints for the amount of tables you can have, that with the current implementation would make it difficult to scale, other backends (such as the proposed SQL backend) will have similar scaling constraints in a multi-tenancy context. 

This layer of abstraction addresses this limitation by allowing us to have greater flexibility to define the schema, and additionally implement our own AuthN/AuthZ layer, and possibly caching too. Whilst we could make changes to the current DynamoDB implementation to allow for multiple installations to share a single DynamoDB table, we then too need to add in an authentication layer (likely using mTLS). We also then need to do this for other backends that Cloud may end up using (PostgreSQL), or even possibly ones in the future that aren't implemented and it's not clear if we would be able to put together a strong argument for them to be implemented if there isn't much demand for it outside of the Cloud team.

For these reasons we (the Cloud team) thought it would be best to shift this responsibility from the Teleport team, onto the Cloud team so we had the freedom to implement a backend that would suit our specific use case, a use case that I don't think has a broad appeal. With that said, implementing an abstraction layer underpinned by gRPC allows external users to implement their own backend to suit their specific needs too.

> How will the GRPC backend persist its data?

Well, this is for the respective backend to decide. For our use case, we would be using DynamoDB for cluster state. It is unlikely that we would use DynamoDB for events however.

**Review by rosstimothy — COMMENTED** [review]
Is there any reason for just adding the proto and stubbing out the backend interface? I would think the actual implementation of this should be pretty straight forward since most of the functions should just be using the grpc client to send the appropriate RPC to the server.

**Comment by michaelmcallister** [comment]
> Is there any reason for just adding the proto and stubbing out the backend interface? I would think the actual implementation of this should be pretty straight forward since most of the functions should just be using the grpc client to send the appropriate RPC to the server.

Purely to keep the change size small and concise. Larger PRs result in more bugs. 

You're right that the implementation is pretty straight forward though, I'm happy to implement these stubs in this PR if you'd prefer

**Comment by michaelmcallister** [comment]
Ping! @zmb3, @rosstimothy

**Review by rosstimothy — APPROVED** [review]

**Comment by zmb3** [comment]
Sorry for letting this slip, @michaelmcallister.

The code changes look fine, but I think we in Teleport Core are still feeling a little bit in the dark as to the plan for the actual backend implementations that will leverage this GRPC interface. We don't want to approve a bunch of your work without knowing what's coming only to have an issue further down the road.

Is the plan that someday we'll support the following backends:

- dynamo
- sqlite
- etcd
- dynamo behind grpc
- sqlite behind grpc
- etcd behind grpc
- ...

Because that kind of feel like where we're headed.

**Review by zmb3 — APPROVED** [review]

**Comment by michaelmcallister** [comment]
@jimbishopp could you please take a look? Still missing some necessary approvals from "set B"

File: lib/backend/grpcclient/grpc.go
Comment (jimbishopp) [inline lib/backend/grpcclient/grpc.go:47]: Maybe we should return an error here until the full implementation is included?

**Review by jimbishopp — APPROVED** [review]
Hi @michaelmcallister. Approving with some thoughts. 

I know you’ve been thinking about this problem for a while now, so forgive me if I’m rehashing things you’ve already discussed multiple times :). No need to respond - just trying to help by sharing my thoughts - don’t want to slow you down.

Some things that come to mind when thinking about this problem:
- I’d first consider adding features to the existing Teleport backends rather than creating new ones. I found out the hard way that backends and caches in Teleport are complex and difficult to implement correctly. It’ll take all of us to support them. And for the same reason, I’m not sure allowing customers to implement their own backend is something we’d want to support, especially for customers with few technical resources.
- I’d be concerned that colocating tenant data in the same Dynamo table is a compliance issue, not to mention a risk for embarrassing and costly bugs. And colocation does not seem to resolve throughput quotas.
- Perhaps a different backend could be used instead of Dynamo - one that doesn’t have the same restrictions. If a new backend is developed, it would be great if it was integrated with Teleport so all customers could take advantage of it. And if an even better backend is created in the future, it would be nice to not be tightly coupled with Dynamo so tenants could be migrated to the shiny new implementation without much effort. I would also assume that we may want the flexibility to use a different backend for different customers depending on the size and expected resource utilization of each customer.
- I’d rather see a complete implementation committed to master instead of a stub. I share your concern for bugs and large PRs, but projects such as this tend to take on a different shape the closer to a full implementation they get. That said, this PR doesn’t look like it is harmful - so it’s up to you - whatever helps you move forward.
- I’m curious about our current run rate - expectation for when we’d hit the Dynamo quotas (no need to discuss here). If it’s really soon, we may want to consider reprioritizing core work to prevent an outage or an inability to onboard new customers.

**Comment by michaelmcallister** [comment]
> I know you’ve been thinking about this problem for a while now, so forgive me if I’m rehashing things you’ve already discussed multiple times :). No need to respond - just trying to help by sharing my thoughts - don’t want to slow you down.

More than happy to document my approach and thoughts, I most definitely don't want to sacrifice documenting things or doing it properly for the sake of speed. 

> I’m not sure allowing customers to implement their own backend is something we’d want to support, especially for customers with few technical resources.

This wouldn't be a good solution for customers with few technical resources, but of course the choice is ultimately up to the developer, and I think it's a good thing to have options. I'm not at all suggesting that external customers/developers write their own gRPC server and have Teleport Core (or Cloud) support it. We'd support the contract (protobuf) and the client only, which I think is a minimal investment on our part.

> I’d be concerned that colocating tenant data in the same Dynamo table is a compliance issue, not to mention a risk for embarrassing and costly bugs. And colocation does not seem to resolve throughput quotas.

Strong isolation is a key requirement for any multi-tenanted solution, and this is no exception. Security is the number one priority. Multiple customer data in same table is no more of a compliance issue than multiple EC2 instances sharing the same underlying hardware, or multiple pods running on a Kubernetes cluster, or customer session data sharing the same S3 bucket. There are technical controls and approaches to ensure strong isolation, but admittedly it is fair to say that it does increase the risk for embarrassing and costly bugs and these will need to be mitigated.

> Perhaps a different backend could be used instead of Dynamo - one that doesn’t have the same restrictions. 

Sure! What did you have in mind? 

A traditional RDBMS solution _could_ scale to possibly low thousands of schemas (one per tenant), though data would still be co-located. There are scalability issues with the amount of concurrent connections but we could address that with solutions such as pgbouncer. For the purposes of cluster state, which is inherently key:value based, an RDBMS wouldn't be a natural fit, but it'd work.

Looking at newer "newSQL" solutions, like CockroachDB, which is an attractive option we can see that they don't really [recommend more than 2,500 tables](https://www.cockroachlabs.com/docs/stable/schema-design-overview.html#quantity-of-tables-and-other-schema-objects). So either multiple clusters would be needed, or a different approach to how isolation is done

>  And colocation does not seem to resolve throughput quotas.

I think it would.

An intermediary like the one proposed could have a caching layer built in, even perhaps plugging in a memcached or redis cluster. Cluster state is read heavy, so an additional layer of caching could alleviate a lot of backend reads to DynamoDB. Whilst Teleport does include a cache, during pod restarts (of which there could be a lot in the case of a host going unavailable) this could prevent the backend being swamped by reads at startup from many pods being started at once.

We use Provisioned throughput, which means upfront we set a floor for each tenant, we then increase this on-demand as necessary. If a tenant is using less than the floor (either it's a quiet period or for whatever reason) then this [quota is still used](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/ServiceQuotas.html) which reduces the available throughput for all tables. With an intermediary in place our provisioned throughput will match closer to our actual usage. With all this said, the throughput quota/limits are more generous than the number of tables per account.

> I’d rather see a complete implementation committed to master instead of a stub. 

If you're happy to review large PRs, I'm happy to write them. I'll update this to include a fuller implementation.

**Comment by michaelmcallister** [comment]
I've implemented the stubs in the latest commit, PTAL @zmb3, @jimbishopp, @rosstimothy, @codingllama 

Outstanding (to be done in subsequent PRs) is:
- Further tests
- online cert rotation for the mTLS auth
- Implement a watcher for the events circular buffer

File: lib/backend/grpcclient/util.go
Comment (rosstimothy) [inline lib/backend/grpcclient/util.go:23]: should this return `nil` if `in` is empty/nil?

File: lib/backend/grpcclient/util.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/util.go:33]: Existing backends only seem to send a nil backend.{Item, Lease, Event} on error, and there are occasions where it'll send an empty response. 

I'm not fussed though, I was just trying to make sure I'm inline with the pre-existing convention

File: lib/backend/grpcclient/util.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/util.go:1]: Totally fair. Went with your suggestion of `proto`

File: lib/backend/grpcclient/util.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/util.go:23]: I don't _think_ so. As before, it seems to me that the implicit expectation is that `nil` be returned only on errors, and for empty responses an empty `backend.Whatever` is returned. I don't mind though, and I may be misinterpreting the expectations of backend implementations

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:184]: Oh good point! No, the results aren't guaranteed to be sorted.

I've added in a sort, thanks for this :)

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:119]: Great idea. I've gone with a 5 second default timeout, but configurable with config.

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:117]: Done. Looks much cleaner, thanks

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go]: I hadn't intended to support this. It would be handy for tests, but even then creating an ephemeral keypair for mTLS in unit tests isn't very hard to do.

Happy to revisit this if there is demand for it though

File: lib/backend/grpcclient/grpc.go
Comment (rosstimothy) [inline lib/backend/grpcclient/grpc.go:163]: Should this be consistent with the etcd backend?

https://github.com/gravitational/teleport/blob/cc2e1ad214037e999f9085b3c28b311230f9d2ff/lib/backend/etcdbk/etcd.go#L276-L277

https://github.com/gravitational/teleport/blob/cc2e1ad214037e999f9085b3c28b311230f9d2ff/api/defaults/defaults.go#L33

File: lib/backend/grpcclient/grpc.go
Comment (rosstimothy) [inline lib/backend/grpcclient/grpc.go:177]:
```suggestion
	return trace.Wrap(b.clientConn.Close())
```

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:163]: Yeah, good idea - thank you!

File: lib/backend/grpcclient/grpc.go
Comment (michaelmcallister) [reply lib/backend/grpcclient/grpc.go:177]: Done

**Comment by michaelmcallister** [comment]
Ping @zmb3 @jimbishopp

**Pull request closed without merging**
